### PR TITLE
[fix][broker] Do not write replicated snapshot marker when the topic which is not enable replication

### PR DESCRIPTION
**Pull request by liangyepianzhou**
### Motivation
[PIP 33](https://github.com/apache/pulsar/wiki/PIP-33%3A-Replicated-subscriptions) introduces a new concept ` Replicated subscriptions`. When a topic has a consumer (subscription) that enables replicated subscriptions, it will write markers into the original topic. Even if there is no replicated cluster configured for this topic, the mark will still be written. And that will make the backlog of the topic keep increasing.

---
The mark will be written in the following two ways:
1. A scheduled task writes a marker at a fixed time interval if there are new messages published.
https://github.com/apache/pulsar/blob/ea1fc0f20138bc35f54f55d32dabf3c3a3309c8e/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/ReplicatedSubscriptionsController.java#L78-L86
https://github.com/apache/pulsar/blob/ea1fc0f20138bc35f54f55d32dabf3c3a3309c8e/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/ReplicatedSubscriptionsSnapshotBuilder.java#L77-L86
2. Acknowledging message will trigger a check if the first snapshot is written and the mark delete position moves,  if true, It will write a marker.
https://github.com/apache/pulsar/blob/ea1fc0f20138bc35f54f55d32dabf3c3a3309c8e/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/ReplicatedSubscriptionsController.java#L114-L150

### Modifications
According to the topic policy to create or remove `ReplicatedSubscriptionsController` of this topic.
### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository:

### CODE REVIEW

**Review by liudezhi2098 — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/ReplicatedSubscriptionsController.java [hunk 1]
```diff
@@ -206,6 +206,12 @@ private void receiveSubscriptionUpdated(ReplicatedSubscriptionsUpdate update) {
     private void startNewSnapshot() {
```
Comment (poorbarcode) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/ReplicatedSubscriptionsController.java:206]: I think the better way is not to create `ReplicatedSubscriptionsController` when turning off `replication.`

Comment (poorbarcode) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/ReplicatedSubscriptionsController.java:206]: I think the better way is not to create `ReplicatedSubscriptionsController` when turning off `replication.`

**Review by poorbarcode — CHANGES_REQUESTED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/ReplicatedSubscriptionsController.java (hunk 1 above)
Comment (liangyepianzhou) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/ReplicatedSubscriptionsController.java:206]: I am trying to use this way for the first time.
But If we do not create `ReplicatedSubscriptionsController ` when turning off `replication`, the `ReplicatedSubscriptionsController` cannot be created when enabling replication at the namespace level.
```
bin/pulsar-admin namespaces set-clusters my-tenant/my-namespace \
--clusters us-west,us-east,us-cent
``` 

Do you have any idea you do this? Adding a new API to start `ReplicatedSubscriptionsController` for all topics in the namespace in all the brokers maybe can do that, but that adds too much complexity.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 2]
```diff
@@ -3476,12 +3476,14 @@ private synchronized void checkReplicatedSubscriptionControllerState(boolean sho
         boolean isCurrentlyEnabled = replicatedSubscriptionsController.isPresent();
         boolean isEnableReplicatedSubscriptions =
                 brokerService.pulsar().getConfiguration().isEnableReplicatedSubscriptions();
+        boolean isReplicationDisEnable = this.topicPolicies.getReplicationClusters().get().size() == 1;
```
Comment (poorbarcode) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:3479]:
```suggestion
        boolean replicationEnabled = this.topicPolicies.getReplicationClusters().get().size() > 1;
```

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ReplicatorSubscriptionTest.java [hunk 3]
```diff
@@ -704,6 +705,53 @@ public void testReplicatedSubscriptionWhenReplicatorProducerIsClosed() throws Ex
         Awaitility.await().untilAsserted(() -> assertNotNull(topic2.getSubscription(subscriptionName)));
     }
 
+    @Test
+    public void testReplicateSubBackLog() throws Exception {
```
Comment (poorbarcode) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ReplicatorSubscriptionTest.java:709]: Please add another test:
- Enable replication
- Pub&Sub messages
- Disabled replication
- Ensure the Replicated Subscription Controller has been closed

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 4]
```diff
@@ -3476,12 +3476,14 @@ private synchronized void checkReplicatedSubscriptionControllerState(boolean sho
...
-        if (shouldBeEnabled && !isCurrentlyEnabled && isEnableReplicatedSubscriptions) {
+        if (shouldBeEnabled && !isCurrentlyEnabled && isEnableReplicatedSubscriptions && !isReplicationDisEnable) {
             log.info("[{}] Enabling replicated subscriptions controller", topic);
             replicatedSubscriptionsController = Optional.of(new ReplicatedSubscriptionsController(this,
                     brokerService.pulsar().getConfiguration().getClusterName()));
-        } else if (isCurrentlyEnabled && !shouldBeEnabled || !isEnableReplicatedSubscriptions) {
+        } else if (isCurrentlyEnabled && !shouldBeEnabled || !isEnableReplicatedSubscriptions
+                || isReplicationDisEnable) {
             log.info("[{}] Disabled replicated subscriptions controller", topic);
             replicatedSubscriptionsController.ifPresent(ReplicatedSubscriptionsController::close);
```
Comment (poorbarcode) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:3488]: At this line, you removed an exist `replicatedSubscriptionsController` instance, right?

If yes, you should stop the timer task `startNewSnapshot` of `replicatedSubscriptionsController` here

And please add a test for this case.

Comment (liangyepianzhou) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:3488]: >If yes, you should stop the timer task startNewSnapshot of replicatedSubscriptionsController here

The logic of `ReplicatedSubscriptionsController::close` is as follow:
```
    @Override
    public void close() {
        timer.cancel(true);
    }
```

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ReplicatorSubscriptionTest.java [hunk 5]
```diff
@@ -704,6 +705,66 @@ public void testReplicatedSubscriptionWhenReplicatorProducerIsClosed() throws Ex
...
+        long backlogSize = topicStats.getBacklogSize();
+        //There should no replicator snapshot marker write into the topic, so the backlog would not change.
+        try {
+            Awaitility.await().untilAsserted(() -> {
+                TopicStats stats = admin1.topics().getStats(topic, false, true);
+                assertNotEquals(stats.getBacklogSize(), backlogSize);
+            });
+            fail();
+        } catch (org.awaitility.core.ConditionTimeoutException ex) {
+        }
```
Comment (poorbarcode) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ReplicatorSubscriptionTest.java:745]: You want to confirm no marker was sent after `10s`, right? It is better to verify after `sleep(10s)`.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ReplicatorSubscriptionTest.java [hunk 6]
```diff
@@ -704,6 +705,66 @@ public void testReplicatedSubscriptionWhenReplicatorProducerIsClosed() throws Ex
...
+                .topic(topic)
+                .subscriptionName(subName)
+                .ackTimeout(5, TimeUnit.SECONDS)
+                .subscriptionType(SubscriptionType.Shared)
+                .replicateSubscriptionState(true)
+                .subscribe();
+        @Cleanup
+        Producer<byte[]> producer = client1.newProducer()
+                .topic(topic)
+                .create();
```
Comment (poorbarcode) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ReplicatorSubscriptionTest.java:731]: Since the replicated replication will publish markers in these scenarios:
- Publish a marker per second if there are new messages are sent
- Ack message that after a snapshot

We'd better make this test flow like this:
- send some messages
- sleep `2s`
- ack messages
- send some messages again
- ack messages

---

And could you also add another test:
- Enable the feature replication on topic level or namespace level
- Pub & Sub
- Disable the feature replication.
- Verify all things work as expected.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ReplicatorSubscriptionTest.java [hunk 7]
```diff
@@ -704,6 +705,66 @@ public void testReplicatedSubscriptionWhenReplicatorProducerIsClosed() throws Ex
...
+            Awaitility.await().untilAsserted(() -> {
+                TopicStats stats = admin1.topics().getStats(topic, false, true);
+                assertNotEquals(stats.getBacklogSize(), backlogSize);
+            });
+            fail();
+        } catch (org.awaitility.core.ConditionTimeoutException ex) {
+        }
+        // Start writing snapshot marker after having remote replicate clusters,
+        // so the backlog would continue to increase.
+        admin1.namespaces().setNamespaceReplicationClusters(namespace, Sets.newHashSet("r1", "r2"));
```
Comment (poorbarcode) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ReplicatorSubscriptionTest.java:748]: Could you also add a case that enabled replication on the topic level?

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ReplicatorSubscriptionTest.java (hunk 6 above)
Comment (liangyepianzhou) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ReplicatorSubscriptionTest.java:740]: Thanks for your suggestion. I have added the ack message case into the test.

**Comment by liangyepianzhou** [comment]
The `DataProvider` of `isTopicPolicyEnabled` will be updated after the [issue](https://github.com/apache/pulsar/pull/21564) is fixed.

File: pulsar-client-api/src/main/java/org/apache/pulsar/client/api/SubscriptionIsolationLevel.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+import org.apache.pulsar.common.classification.InterfaceStability;
+
+/**
+ * When creating a consumer, if the subscription does not exist, a new subscription will be created.
+ * The default isolation level for Subscription is 'READ_COMMITTED'.
+ * See {@link #subscriptionIsolationLevel(SubscriptionIsolationLevel)} to configure the isolation level behavior.
+ */
+@InterfaceAudience.Public
+@InterfaceStability.Stable
+public enum SubscriptionIsolationLevel {
```
Comment (poorbarcode) [inline pulsar-client-api/src/main/java/org/apache/pulsar/client/api/SubscriptionIsolationLevel.java:31]: This class is not related to current PR, right?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 9]
```diff
@@ -3010,7 +3010,7 @@ public CompletableFuture<Void> onPoliciesUpdate(@Nonnull Policies data) {
         }
 
         updateTopicPolicyByNamespacePolicy(data);
-
+        checkReplicatedSubscriptionControllerState();
```
Comment (poorbarcode) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:3013]: If there is no subscription that enabled `replicatedSubscription`, we should not start the scheduled snapshot task, right?

File: pulsar-client-api/src/main/java/org/apache/pulsar/client/api/SubscriptionIsolationLevel.java (hunk 8 above)
Comment (liangyepianzhou) [reply pulsar-client-api/src/main/java/org/apache/pulsar/client/api/SubscriptionIsolationLevel.java:31]: Yeah, It should be introduced by merging the master branch.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 9 above)
Comment (liangyepianzhou) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:3013]: https://github.com/apache/pulsar/blob/62afe0194e0636be2a0468fa7f47f271442ba0c8/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java#L3479-L3494

**Review by poorbarcode — APPROVED** [review]

**Comment by lhotari** [comment]
Please also review #16651. That's a long out-standing PR that fixes issues with replicated snapshot marker messages.
